massa-execution-worker/src/speculative_roll_state.rs

```diff
@@ -118,34 +118,48 @@ impl SpeculativeRollState {
             )));
         }
 
-        let cur_cycle = slot.get_cycle(periods_per_cycle);
-
-        // remove the rolls
-        let current_rolls = self
-            .added_changes
-            .roll_changes
-            .entry(*seller_addr)
-            .or_insert_with(|| owned_count);
-        *current_rolls = owned_count.saturating_sub(roll_count);
-
         // compute deferred credit slot
+        let cur_cycle = slot.get_cycle(periods_per_cycle);
         let target_slot = Slot::new_last_of_cycle(
             cur_cycle
                 .checked_add(3)
                 .expect("unexpected cycle overflow in try_sell_rolls"),
             periods_per_cycle,
             thread_count,
         )
-        .expect("unexepected slot overflot in try_sell_rolls");
+        .expect("unexpected slot overflow in try_sell_rolls");
+
+        // Note 1: Deferred credits are stored as absolute value
+        // Note 2: Deferred credits are credited at the beginning of the slot execution
+        // (before the block is executed)
+        // So we retrieve deferred credits from the next slot
+        let next_slot = slot.get_next_slot(slot.thread)
+            .expect("Unable to compute next slot");
+        let deferred_credits = self.get_address_deferred_credits(&seller_addr, next_slot);
+        let deferred_credits_amount = deferred_credits
+            .into_iter()
+            .last()
+            .map(|(_slot, amount)| amount)
+            .unwrap_or_default();
```

Comment:
This effectively retrieves all the registered future credits for that address, and takes the last one.
This is not the expected behavior.
We are only looking for the deferred credits currently registered for target_slot.

Here is what I propose:

In deferred_credits.rs / impl DeferredCredits:

```rust
// Gets the deferred credits for a given address that will be credited at a given slot
fn get_address_deferred_credit_for_slot(&self, addr: &Address, slot: &Slot) -> Option<Amount> {
    if let Some(v) = self.0.get(slot).and_then(|slot_credits| slot_credits.get(addr)) {
        return Some(v); 
    }
    None
}
```



In active_history:


```rust
// Gets the deferred credits for a given address that will be credited at a given slot
fn get_address_deferred_credit_for_slot(&self, addr: &Address, slot: &Slot) -> Option<Amount> {
    for hist_item in self.0.iter().rev() {
        if let Some(v) = hist_item.state_changes.pos_changes.deferred_credits.get_address_deferred_credit_for_slot(addr, slot) {
            return Some(v);     
        }
    }
    None
}
```


In pos_final_state.rs / impl PoSFinalState:


```rust
// Gets the deferred credits for a given address that will be credited at a given slot
fn get_address_deferred_credit_for_slot(&self, addr: &Address, slot: &Slot) -> Option<Amount> {
    if let Some(v) = self.deferred_credits.get_address_deferred_credit_for_slot(addr, slot) {
        return Some(v); 
    }
}
```

In speculative_roll_state:


```rust
// Gets the deferred credits for a given address that will be credited at a given slot
fn get_address_deferred_credit_for_slot(&self, addr: &Address, slot: &Slot) -> Option<Amount> {
    
    // search in the added changes
    if let Some(v) = self.added_changes.deferred_credits.get_address_deferred_credit_for_slot(addr, slot) {
        return Some(v); 
    }
    
    // search in the history
    if let Some(v) = self.active_history.read().get_address_deferred_credit_for_slot(addr, slot).copied() {
        return Some(v);
    }
    
    // search in the final state
    if let Some(v) = self.final_state.read().pos_state.get_address_deferred_credit_for_slot(addr, slot).copied() {
        return Some(v);
    }
    
    None
}
```

Then, here you can just do 
```rust
let deferred_credits = get_address_deferred_credit_for_slot(&seller_addr, &target_slot).unwrap_or_default();
```

Reply:
done